src/v/cluster/producer_state.cc

```diff
@@ -0,0 +1,264 @@
+      || (last_req && last_req.value()->_last_sequence == std::numeric_limits<seq_t>::max() && incoming == 0);
+}
+
+result<request_ptr> requests::try_emplace(
+  seq_t first, seq_t last, model::term_id current, bool reset) {
+    if (reset) {
+        // reset all the sequence tracking state, avoids any sequence
+        // checks for sequence tracking.
+        while (!_inflight_requests.empty()) {
+            if (!_inflight_requests.front()->has_completed()) {
+                _inflight_requests.front()->set_value(errc::timeout);
+            }
+            _inflight_requests.pop_front();
+        }
+        _finished_requests.clear();
+    } else {
+        // gc and fail any inflight requests from old terms
+        // these are guaranteed to be failed because of sync() guarantees
+        // prior to this request.
+        while (!_inflight_requests.empty()
+               && _inflight_requests.front()->_term < current) {
+            if (!_inflight_requests.front()->has_completed()) {
+                // Here we know for sure the term change, these in flight
+                // requests are going to fail anyway, mark them so.
+                _inflight_requests.front()->set_value(errc::timeout);
+            }
+            _inflight_requests.pop_front();
+        }
+
+        // check if an existing request matches
+        auto match_it = std::find_if(
+          _finished_requests.begin(),
+          _finished_requests.end(),
+          [first, last](const auto& request) {
+              return request->_first_sequence == first
+                     && request->_last_sequence == last;
+          });
+
+        if (match_it != _finished_requests.end()) {
+            return *match_it;
+        }
+
+        match_it = std::find_if(
+          _inflight_requests.begin(),
+          _inflight_requests.end(),
+          [first, last, current](const auto& request) {
+              return request->_first_sequence == first
+                     && request->_last_sequence == last
+                     && request->_term == current;
+          });
+
+        if (match_it != _inflight_requests.end()) {
+            return *match_it;
+        }
+
+        if (!is_valid_sequence(first)) {
+            vlog(
+              clusterlog.trace,
+              "out of order sequence for request range [{}, {}], term: {}",
+              first,
+              last,
+              current);
+            return errc::sequence_out_of_order;
+        }
+    }
+
+    // All invariants satisfied, enqueue the request.
+    _inflight_requests.emplace_back(
+      ss::make_lw_shared<request>(first, last, current, result_promise_t{}));
+
+    return _inflight_requests.back();
+}
+
+bool requests::stm_apply(
+  const model::batch_identity& bid, kafka::offset offset) {
+    bool relink_producer = false;
+    auto first = bid.first_seq;
+    auto last = bid.last_seq;
+    if (!_inflight_requests.empty()) {
```

Comment:
when two branches are present it's better to avoid negation operator by swapping branches